k2/python/k2/mwer_loss.py

```diff
@@ -0,0 +1,140 @@
+          use_double_scores:
+            True to use double precision floating point.
+            False to use single precision.
+        '''
+
+        super().__init__()
+        self.temperature = temperature
+        self.use_double_scores = use_double_scores
+
+    def forward(self,
+                lattice: k2.Fsa,
+                ref_texts: Union[k2.RaggedTensor, List[List[int]]],
+                nbest_scale: float,
+                num_paths: int) -> torch.Tensor:
+        '''Compute the Minimum Word Error loss given
+        a lattice and corresponding ref_texts.
+
+        Args:
+          lattice:
+            An FsaVec with axes [utt][state][arc].
+          ref_texts:
+            It can be one of the following types:
+              - A list of list-of-integers, e..g, `[ [1, 2], [1, 2, 3] ]`
+              - An instance of :class:`k2.RaggedTensor`.
+                Must have `num_axes == 2` and with dtype `torch.int32`.
+          nbest_scale:
+            Scale `lattice.score` before passing it to :func:`k2.random_paths`.
+            A smaller value leads to more unique paths at the risk of being not
+            to sample the path with the best score.
+          num_paths:
+            Number of paths to **sample** from the lattice
+            using :func:`k2.random_paths`.
+        Returns:
+            Minimum Word Error Rate loss.
+        '''
+
+        nbest = k2.Nbest.from_lattice(
+            lattice=lattice,
+            num_paths=num_paths,
+            use_double_scores=self.use_double_scores,
+            nbest_scale=nbest_scale,
+        )
+        device = lattice.scores.device
+        path_arc_shape = nbest.kept_path.shape.to(device)
+        stream_path_shape = nbest.shape.to(device)
+
+        hyps = nbest.build_levenshtein_graphs()
+        refs = k2.levenshtein_graph(ref_texts, device=hyps.device)
+        levenshtein_alignment = k2.levenshtein_alignment(
+            refs=refs,
+            hyps=hyps,
+            hyp_to_ref_map=nbest.shape.row_ids(1),
+            sorted_match_ref=True,
+        )
+        tot_scores = levenshtein_alignment.get_tot_scores(
+            use_double_scores=self.use_double_scores, log_semiring=False
+        )
+        # Each path has a corresponding wer.
+        wers = -tot_scores.to(device)
+
+        # Group each log_prob into [path][arc]
+        ragged_nbest_logp = k2.RaggedTensor(path_arc_shape, nbest.fsa.scores)
+        # Get the probalitity of each path, in log format,
+        # with shape [stream][path].
+        path_logp = ragged_nbest_logp.sum() / self.temperature
+        ragged_path_prob = k2.RaggedTensor(stream_path_shape, path_logp.exp())
+
+        # Normalize prob of each path.
+        den_prob = ragged_path_prob.sum()
+        den_logp = torch.index_select(
+            den_prob.log(), 0, ragged_path_prob.shape.row_ids(1))
+        prob_normalized = (path_logp - den_logp).exp()
+
+        loss = (prob_normalized * wers).sum()
```

Comment:
Hi k2 team,
Great PR!

I'm not sure what size `prob_normalized` and `wers` have here.
If their length is `num_utterances` (batch size), then wouldn't it be better to return the loss as is (without summation)? In particular, so that the user can apply `.sum()` or `.mean()` to the final loss (e.g. after adding mwer loss to the `.get_tot_scores()` result of the main loss).
If their length is `num_utterances * num_paths` then does it make sense to do summation with respect to the utterance number, to, again, get the loss size as `num_utterances`?

Reply:
The length is not num_utterance. While it's  not num_utterances * num_paths either.  Since num_paths of each utterance could be unequal.  
Its size is `num_paths_of_utt1 + num_paths_of_utt2 + ... + num_paths_of_uttN`.

Maybe there should be a `reduction` option. 
If `none` then ` return the loss as is (without summation)` as you suggested.
If `sum` then return a summed value as what it does now.
If `mean` return a mean value.

Do you have suggestions?  @GNroy

Reply:
@glynpu thanks for answering!

I believe that there should be an internal reduction option of loss values belonging to the same utterance. In particular, get one number as the sum or mean over all paths of `utt1`, then the same for `utt2`, and so on.
I find it quite intuitive to get `n` loss values when feeding `n` utterances (lattices). Also, having loss values per utterance (instead of one per batch) enables tricks like post scaling (e.g. focal loss).

Regular `reduction` is also an option for those who just want a single loss value.

Reply:
I agree about the reducction option.

Reply:
What do you think of current design? @GNroy @danpovey 

> I find it quite intuitive to get n loss values when feeding n utterances (lattices). Also, having loss values per utterance (instead of one per batch) enables tricks like post scaling (e.g. focal loss).

This requirement is quite easy, you just need one extra step to get loss values per utterance. 
An example usage could be
```
# loss is a k2.RaggedTensor as explained in the comment.
loss = k2.mwer_loss(lattice, refs_texts, reduction='none')
loss_per_utt = loss.sum()
```

https://github.com/k2-fsa/k2/blob/cc67ec0f4622dadc1467f7e9cd040e73015ef478/k2/python/k2/mwer_loss.py#L160-L172

Reply:
LGTM except for a small typo on lines 43 and 172: tot**o**l -> tot**a**l
Thanks!